### PR TITLE
In-memory engine: fix unsafe-destroy-range/delete-range does not trigger eviction

### LINKED ISSUE #17644: In-memory Engine: GC worker delete range does not evict relevant regions
## Bug Report

After https://github.com/tikv/tikv/pull/17359, delete range in GC worker no longer evict relevant regions, because it calls `delete_range` directly from the RocksDB instance, which is not observed by `WriteBatchWrapper`.

### What version of TiKV are you using?

v8.4.0

### What did you expect?

GC worker delete range should evict relevant regions.

### What did happened?

GC worker delete range does not evict relevant regions.

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Close #17644 Ref #16141

What's Changed:

```commit-message
fix unsafe-destroy-range/delete-range does not trigger eviction
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/hybrid_engine/src/observer/load_eviction.rs [hunk 1]
```diff
@@ -133,6 +133,31 @@ impl LoadEvictionObserver {
...
 
 impl QueryObserver for LoadEvictionObserver {
+    fn pre_exec_query(
+        &self,
+        ctx: &mut ObserverContext<'_>,
+        reqs: &[kvproto::raft_cmdpb::Request],
+        _: u64,
+        _: u64,
+    ) -> bool {
+        reqs.iter().for_each(|r| {
```
Comment (overvenus) [inline components/hybrid_engine/src/observer/load_eviction.rs:143]:
```suggestion
        for r in reqs {
```

Prefer for loop for better readability.

File: src/server/gc_worker/gc_worker.rs [hunk 2]
```diff
@@ -192,6 +192,8 @@ pub struct GcRunnerCore<E: Engine> {
     cfg_tracker: Tracker<GcConfig>,
 
     stats_map: HashMap<GcKeyMode, Statistics>,
+
+    coprocessor_hook: CoprocessorHost<E::Local>,
```
Comment (overvenus) [inline src/server/gc_worker/gc_worker.rs:196]:
```suggestion
    coprocessor_host: CoprocessorHost<E::Local>,
```

File: components/raftstore/src/coprocessor/dispatcher.rs [hunk 3]
```diff
@@ -616,6 +616,21 @@ impl<E: KvEngine> CoprocessorHost<E> {
...
 
+    pub fn pre_delete_range(&self, start_key: &[u8], end_key: &[u8]) {
+        let region = Region::default();
+        let mut ctx = ObserverContext::new(&region);
+        for observer in &self.registry.query_observers {
+            let observer = observer.observer.inner();
+            let mut request = Request::new();
+            request.set_cmd_type(CmdType::DeleteRange);
+            request.mut_delete_range().set_start_key(start_key.to_vec());
+            request.mut_delete_range().set_end_key(end_key.to_vec());
```
Comment (overvenus) [inline components/raftstore/src/coprocessor/dispatcher.rs:627]: To maintain consistency in the request, it's best not to include data keys (with a 'z' prefix) in the request. In this pull request, the caller (`GcRunnerCore`) is passing data keys, which is not recommended.

File: src/server/gc_worker/gc_worker.rs [hunk 4]
```diff
@@ -747,6 +752,9 @@ impl<E: Engine> GcRunnerCore<E> {
             let start_data_key = keys::data_key(start_key.as_encoded());
             let end_data_key = keys::data_end_key(end_key.as_encoded());
 
+            self.coprocessor_hook
+                .pre_delete_range(&start_data_key, &end_data_key);
+
```
Comment (overvenus) [inline src/server/gc_worker/gc_worker.rs:757]:
```suggestion
            self.coprocessor_hook
                .pre_delete_range(start_key.as_encoded(), end_key.as_encoded());
            let start_data_key = keys::data_key(start_key.as_encoded());
            let end_data_key = keys::data_end_key(end_key.as_encoded());

```

File: components/hybrid_engine/src/observer/load_eviction.rs (hunk 1 above)
Comment (glorv) [reply components/hybrid_engine/src/observer/load_eviction.rs:143]: I suppose rust should prefer functional than procedural.

**Review by glorv — APPROVED** [review]

File: src/server/gc_worker/gc_worker.rs (hunk 4 above)
Comment (glorv) [reply src/server/gc_worker/gc_worker.rs:757]: delete_range in write batch also shares the DeleteRange event in IME, so I think we should keep consistency between them. And `delete_range` in write_batch uses data_key

File: components/hybrid_engine/src/observer/load_eviction.rs (hunk 1 above)
Comment (SpadeA-Tang) [reply components/hybrid_engine/src/observer/load_eviction.rs:143]: +1

File: tests/failpoints/cases/test_region_cache_engine.rs [hunk 5]
```diff
@@ -39,10 +50,40 @@ fn must_copr_point_get(cluster: &mut Cluster<ServerCluster>, table: &ProductTabl
         .key_ranges(vec![key_range])
         .build_with(ctx, &[0]);
```
Comment (overvenus) [inline tests/failpoints/cases/test_region_cache_engine.rs:51]:
```suggestion
        .key_ranges(vec![key_range])
        .start_ts(get_tso(&cluster.pd_client).into())
        .build_with(ctx, &[0]);
```

I forgot to add the latest TSO.

**Review by overvenus — APPROVED** [review]

**Pull request merged**